File: rs/anda_cognitive_nexus/src/projection/dependency.rs
//! Read-time dependency validity. Stored cognition is never rewritten by this check.
use super::Policy;
use crate::{
    id::ElementId,
    kql::Context,
    store::{Element, rows::state},
};
use anda_kip::{ElementKind, Json, KipError};

#[derive(Default)]
struct Validity {
    // 0=current, 1=needs_review, 2=unverifiable. Reasons reveal no hidden refs/counts.
    state: u8,
    reasons: Vec<String>,
    next: Option<String>,
}
impl Validity {
    fn issue(state: u8, reason: &str) -> Self {
        Self {
            state,
            reasons: vec![reason.into()],
            next: None,
        }
    }
    fn absorb(&mut self, other: Self) {
        self.state = self.state.max(other.state);
        self.reasons.extend(other.reasons);
        self.next = self.next.iter().chain(other.next.iter()).min().cloned();
    }
}

impl Context<'_> {
    pub(crate) async fn dependency_validity(
        &mut self,
        element: &Element,
        policy: &Policy,
        at: &str,
    ) -> Result<Json, KipError> {
        if policy.trust_version == "unavailable" {
            return Ok(
                serde_json::json!({"status":"unverifiable","action_eligible":false,"reasons":["historical projection control state unavailable"],"basis":self.projection_basis(policy,at,None)}),
            );
        }
        let cache_key = crate::schema::contracts::digest(&serde_json::json!([
            element.id().to_string(),
            element.version(),
            self.pinned_seq,
            policy,
            at,
        ]))?;
        if let Some(value) = self.dependency_cache.get(&cache_key) {
            return Ok(value.clone());
        }
        let mut result = self
            .check_dependencies(element, policy, at, &mut Vec::new())
            .await?;
        result.reasons.sort();
        result.reasons.dedup();
        let value = serde_json::json!({
            "status": (["current", "needs_review", "unverifiable"][result.state as usize]),
            "action_eligible": result.state == 0,
            "reasons": result.reasons,
            "basis": self.projection_basis(policy, at, result.next),
        });
        self.dependency_cache.insert(cache_key, value.clone());
        Ok(value)
    }

    async fn check_dependencies(
        &mut self,
        element: &Element,
        policy: &Policy,
        at: &str,
        path: &mut Vec<ElementId>,
    ) -> Result<Validity, KipError> {
        let id = element.id();
        if path.contains(&id) || path.len() >= 64 {
            return Ok(Validity::issue(2, "dependency cycle or traversal limit"));
        }
        let Some(visibility) = self
            .authority
            .may_read(element, self.auth)
            .filter(|v| v.content)
        else {
            return Ok(Validity::issue(2, "dependency source unavailable"));
        };
        // A masked source cannot establish that the actor may inspect all of a
        // pin's prerequisites. Do not turn masked fields into a positive check.
        if !visibility.constraints.fields.is_empty() {
            return Ok(Validity::issue(2, "dependency source is field-restricted"));
        }
        if element.state() != state::ACTIVE && element.state() != state::MERGED {
            return Ok(Validity::issue(1, "dependency lifecycle changed"));
        }
        let mut result = Validity::default();
        match element {
            Element::Evidence(row) if row.status == "corrected" => {
                return Ok(Validity::issue(1, "dependency evidence corrected"));
            }
            Element::Assertion(row) => {
                // Its own written interval, with a missing `from` read as
                // {latest: asserted_at} (§25.2); `expired` is computed (§14.3).
                let timed = super::world::Timed::new(row, id, String::new());
                if !matches!(row.status.as_str(), "active" | "expired")
                    || timed.place(at) == super::world::Placement::Outside
                {
                    return Ok(Validity::issue(1, "dependency no longer eligible"));
                }
                result.next = timed.boundaries_after(at).min().map(str::to_string);
            }
            _ => {}
        }
        if self
            .store
            .control_at(
                &self.space,
                &format!("identity_review/{id}"),
                self.pinned_seq,
            )
            .await?
            .is_some()
        {
            return Ok(Validity::issue(
                1,
                "identity interpretation requires review",
            ));
        }
        if !crate::schema::contracts::is_derived(element) {
            return Ok(result);
        }
        let mut producer = None;
        let ids = self
            .candidates(
                ElementKind::Activity,
                Some(crate::store::eq_fields(&[
                    ("space", anda_db_schema::Fv::Text(self.space.clone())),
                    (
                        "output_keys",
                        anda_db_schema::Fv::Text(crate::term::Endpoint::Local(id).key()),
                    ),
                ])),
            )
            .await?;
        self.charge(ids.len())?;
        for aid in ids {
            let Some(Element::Activity(row)) = self.load_unattached(aid).await? else {
                continue;
            };
            if !self
                .authority
                .may_read(&Element::Activity(row.clone()), self.auth)
                .is_some_and(|v| v.content)
            {
                continue;
            }
            if row.activity_class != "dependency_validation"
                && row.created_tx != *element.envelope().created_tx
                && row.updated_tx != *element.envelope().updated_tx
            {
                continue; // A retrospective audit is not the producing computation.
            }
            if row.status != "completed" || row.state != state::ACTIVE {
                continue;
            }
            if row.origin["_kip_runtime"]["output_versions"][id.to_string()].as_u64()
                != Some(element.version())
            {
                continue;
            }
            if !row
                .outputs
                .iter()
                .any(|r| r.as_str().or_else(|| r["id"].as_str()) == Some(id.to_string().as_str()))
            {
                continue;
            }
            if let Some((_, basis)) = row
                .facets
                .iter()
                .find(|(name, _)| name.ends_with("/DependencyBasis"))
                && producer.as_ref().is_none_or(|(seq, _)| row.seq > *seq)
            {
                producer = Some((row.seq, basis.clone()));
            }
        }
        let Some((_, contract)) = producer else {
            return Ok(Validity::issue(
                2,
                "exact producing DependencyBasis unavailable",
            ));
        };
        if crate::schema::contracts::validate_value(&serde_json::json!({"$ref":"urn:kip:2.0:schema:cognitive-records#/$defs/DependencyBasis"}), &contract).is_err() {
            return Ok(Validity::issue(2, "dependency contract is invalid"));
        }
        let basis = serde_json::to_value(self.projection_basis(policy, at, None)).unwrap();
        for coordinate in [
            "schema_environment_version",
            "identity_version",
            "policy",
            "trust_version",
            "purpose",
            "risk",
        ] {
            if contract["policy_basis"][coordinate] != basis[coordinate] {
                result.absorb(Validity::issue(1, "dependency computation policy changed"));
            }
        }
        if let Some(contexts) = contract["policy_basis"]["context_refs"].as_array()
            && !contexts.iter().all(|c| {
                basis["context_refs"]
                    .as_array()
                    .is_some_and(|refs| refs.contains(c))
            })
        {
            result.absorb(Validity::issue(1, "dependency context mismatch"));
        }
        path.push(id);
        for group in contract["groups"].as_array().into_iter().flatten() {
            let mut members = Vec::new();
            for pin in group["pins"].as_array().into_iter().flatten() {
                let source = pin["id"]
                    .as_str()
                    .and_then(|id| id.parse::<ElementId>().ok());
                let Some(source) = source else {
                    members.push(Validity::issue(2, "dependency source unavailable"));
                    continue;
                };
                let Some(row) = self.load_unattached(source).await? else {
                    members.push(Validity::issue(2, "dependency source unavailable"));
                    continue;
                };
                if !self
                    .authority
                    .may_read(&row, self.auth)
                    .is_some_and(|v| v.content)
                {
                    members.push(Validity::issue(2, "dependency source unavailable"));
                    continue;
                }
                let pinned_planes = pin["planes"].as_object().filter(|v| !v.is_empty());
                let changed = if let Some(planes) = pinned_planes {
                    let current = serde_json::to_value(row.plane_versions()).unwrap();
                    planes.iter().any(|(name, value)| {
                        crate::schema::contracts::pinned_plane(&current, name) != value.as_u64()
                    })
                } else {
                    pin["version"].as_u64() != Some(row.version())
                };
                if changed {
                    members.push(Validity::issue(1, "dependency source version changed"));
                } else {
                    members.push(Box::pin(self.check_dependencies(&row, policy, at, path)).await?);
                }
            }
            match group["role"].as_str() {
                Some("context") => {
                    if members.iter().any(|m| m.state != 0) {
                        result.reasons.push("context dependency changed".into());
                    }
                }
                Some("any_of") if members.iter().any(|m| m.state == 0) => {
                    for member in members.into_iter().filter(|m| m.state == 0) {
                        result.absorb(member);
                    }
                }
                _ => {
                    for member in members {
                        result.absorb(member);
                    }
                }
            }
        }
        path.pop();
        Ok(result)
    }
}

File: ts/kip-do/src/projection/dependency.ts
/** Cognitive Consistency §3: computed validity, never an author-written flag. */
import type { Context } from '../kql/context.js'
import { isInvalidated } from '../recording.js'
import { tryParseElementId, formatElementId } from '../id.js'
import { canonicalJson, isJsonMap, type JsonMap } from '../json.js'
import { State, type ActivityRow, type Element } from '../store/index.js'
import { validateValue, pinnedPlane } from '../schema/contracts.js'
import { projectionBasis } from './index.js'
import type { Policy } from './policy.js'

export function isDerived(element: Element): boolean {
  if (element.kind === 'Assertion') return element.row.mode === 'inferred'
  if (element.kind !== 'Concept') return false
  return ['SkillRevision', 'Insight', 'WorkingState'].some((name) => element.row.schema_ref.endsWith('/' + name)) ||
    Object.keys(element.row.structural).some((name) => name.endsWith('/derived_from'))
}
interface Validity { state: number; reasons: string[]; next: string | null }
const issue = (state: number, reason: string): Validity => ({ state, reasons: [reason], next: null })
const absorb = (into: Validity, other: Validity): void => {
  into.state = Math.max(into.state, other.state)
  into.reasons.push(...other.reasons)
  into.next = [into.next, other.next].filter((v): v is string => v !== null).sort()[0] ?? null
}

export function dependencyValidity(cx: Context, element: Element, policy: Policy, at: string): JsonMap {
  if (!policy.trust_version || policy.trust_version === 'unavailable') return {status:'unverifiable',action_eligible:false,reasons:['historical projection control state unavailable'],basis:projectionBasis(cx,policy,at)}
  function check(element: Element, path: string[]): Validity {
    const id = formatElementId({ kind: element.kind, seq: element.row.id })
    if (path.includes(id) || path.length >= 64) return issue(2, 'dependency cycle or traversal limit')
    const visibility = cx.authority.mayRead(element, cx.auth)
    if (!visibility?.content || visibility.constraints.fields.length) return issue(2, 'dependency source unavailable')
    if (element.row.state !== State.ACTIVE && element.row.state !== State.MERGED) return issue(1, 'dependency lifecycle changed')
    if (isInvalidated(element)) return issue(1, 'dependency extraction repaired')
    const result: Validity = { state: 0, reasons: [], next: null }
    if (element.kind === 'Evidence' && element.row.status === 'corrected') return issue(1, 'dependency evidence corrected')
    if (element.kind === 'Assertion') {
      const row = element.row
      if (row.status !== 'active' || (row.valid_from && row.valid_from > at) || (row.valid_until && row.valid_until <= at)) return issue(1, 'dependency no longer eligible')
      result.next = [row.valid_from, row.valid_until].filter((t) => t > at).sort()[0] ?? null
    }
    if (cx.store.controlAt(cx.space, `identity_review/${id}`, cx.asOf ?? cx.store.currentSeq(cx.space))) return issue(1, 'identity interpretation requires review')
    if (!isDerived(element)) return result
    const activities = cx.asOf !== null
      ? cx.reconstruct('Activity').map((e) => e.row as ActivityRow)
      : cx.store.all<ActivityRow>('activities', 'SELECT * FROM activities WHERE space = ? ORDER BY id', cx.space)
    cx.spend('scans', activities.length)
    let producer: { seq: number; contract: JsonMap } | null = null
    for (const activity of activities) {
      const readable = cx.load({ kind: 'Activity', seq: activity.id }, false)
      if (!readable || !cx.authority.mayRead(readable, cx.auth)?.content) continue
      if (activity.activity_class !== 'dependency_validation' && activity.created_tx !== element.row.created_tx && activity.updated_tx !== element.row.updated_tx) continue
      if (activity.status !== 'completed' || activity.state !== State.ACTIVE) continue
      const runtime = activity.origin._kip_runtime as JsonMap | undefined
      if (!runtime || (runtime.output_versions as JsonMap)?.[id] !== element.row.version) continue
      if (!activity.outputs.some((ref) => (typeof ref === 'string' ? ref : (ref as JsonMap).id) === id)) continue
      const contract = Object.entries(activity.facets).find(([name]) => name.endsWith('/DependencyBasis'))?.[1]
      if (isJsonMap(contract) && (!producer || activity.seq > producer.seq)) producer = { seq: activity.seq, contract }
    }
    if (!producer) return issue(2, 'exact producing DependencyBasis unavailable')
    const { contract } = producer
    try { validateValue({ $ref: 'urn:kip:2.0:schema:cognitive-records#/$defs/DependencyBasis' }, contract) }
    catch { return issue(2, 'dependency contract is invalid') }
    const basis = projectionBasis(cx, policy, at)
    const old = contract.policy_basis as JsonMap
    for (const coordinate of ['schema_environment_version', 'identity_version', 'policy', 'trust_version', 'purpose', 'risk']) {
      if (canonicalJson(old[coordinate]) !== canonicalJson(basis[coordinate])) absorb(result, issue(1, 'dependency computation policy changed'))
    }
    if (!(old.context_refs as string[]).every((r) => (basis.context_refs as string[]).includes(r))) absorb(result, issue(1, 'dependency context mismatch'))
    for (const group of contract.groups as JsonMap[]) {
      const members = (group.pins as JsonMap[]).map((pin): Validity => {
        const source = tryParseElementId(String(pin.id))
        const row = source && cx.load(source, false)
        if (!row || !cx.authority.mayRead(row, cx.auth)?.content) return issue(2, 'dependency source unavailable')
        const planes = isJsonMap(pin.planes) && Object.keys(pin.planes).length ? pin.planes : null
        const changed = planes ? Object.entries(planes).some(([name, version]) => pinnedPlane(row.row.plane_versions as unknown as JsonMap, name) !== version) : row.row.version !== pin.version
        return changed ? issue(1, 'dependency source version changed') : check(row, [...path, id])
      })
      if (group.role === 'context') {
        if (members.some((m) => m.state)) result.reasons.push('context dependency changed')
      } else if (group.role === 'any_of' && members.some((m) => m.state === 0)) {
        members.filter((m) => m.state === 0).forEach((m) => absorb(result, m))
      } else members.forEach((m) => absorb(result, m))
    }
    return result
  }
  const checked = check(element, [])
  return { status: ['current', 'needs_review', 'unverifiable'][checked.state]!, action_eligible: checked.state === 0, reasons: [...new Set(checked.reasons)].sort(), basis: projectionBasis(cx, policy, at, checked.next) }
}
